### ISSUE TITLE
Bokeh Support for Keras

### ISSUE THREAD

**Issue by AntreasAntoniou** [fchollet/keras#1101]
Dear Fchollet,

Thanks for the amazing tool you made for us. I was wondering if you would be interested in adding support for the open source visualisation library bokeh for keras. I have made a small library that allows one to create a graph with a single line and then update it with one more line when a new data point appears. Using the bokeh server you can have real time interactive charts. Would you be interested in adding this to your library ?

**Comment by fchollet** [comment]
> Would you be interested in adding this to your library ?

In general I believe that visualization tools would benefit from being developed separately from Keras, and integrated into Keras via the use of callbacks.

If you have a repo with visualization tools for Keras built on top of Bokeh, I'd be happy to check it out.

**Comment by AntreasAntoniou** [comment]
I will make it into a nice repo and give you a link soon enough

**Comment by stared** [comment]
For interactive charts, see this gist: [Live loss plots in Jupyter Notebook for Keras](https://gist.github.com/stared/dfb4dfaf6d9a8501cd1cc8b8cb806d2e). It uses plain `matplotlib` (rather thank Bokeh), via callbacks.

@fchollet  After suitable rewriting, would you like to have it in `callbacks`?

@AntreasAntoniou By any chance, did you make any progress on the Bokeh plot for Keras?

**Comment by fchollet** [comment]
> After suitable rewriting, would you like to have it in callbacks?

Possibly. A callback for live Jupyter plot is definitely a nice feature to have. But it seems to me that the API should be more general than just loss plotting (what about metrics, like accuracy, etc?). And if we plot more than one metric type, we need to think about the scale issue: different quantities will have different scales (e.g. loss vs. accuracy) and will require separate plots. Etc... a few questions to answer.

**Comment by stared** [comment]
Great!

By rewriting I meant all of these things, not just some refactoring. My biggest reservation here is if I can make it play with Keras progress bar.

**Comment by Abhijit-2592** [comment]
@fchollet  I have modified @stared 's code to make the dynamic plots much more interactive. I have plotted all (loss, acc , val_loss , val_acc) in a single plot using bokeh. I have addressed the problem you stated " And if we plot more than one metric type, we need to think about the scale issue: different quantities will have different scales (e.g. loss vs. accuracy) and will require separate plots." by making the plot interactive. If you like it add it to keras call backs.
Please look at the code [here](https://github.com/Abhijit-2592/Keras-custom-callbacks)

**Comment by stared** [comment]
@Abhijit-2592  Cool! Though, accuracy and log-loss shouldn't be on the same plot, as they use different units.

Compare with this one: https://github.com/deepsense-ai/intel-ai-webinar-neural-networks/blob/master/notebook_done.ipynb

![download 3](https://user-images.githubusercontent.com/1001610/28584693-ea6c421c-716d-11e7-81dc-33e105b072cd.png)

Just code here: https://github.com/deepsense-ai/intel-ai-webinar-neural-networks/blob/master/live_loss_plot.py

**Comment by Abhijit-2592** [comment]
@stared , Thanks for the input. That notebook was a rough draft. will work this weekend and get back with a proper visualisation function. And have you found a way to make it work with keras progress bar? I am working on that too. I will let you know once I figure it out :)

**Comment by stared** [comment]
EDIT: I created a `livelossplot` package - with a Keras callback and a generic API, see here: https://github.com/stared/livelossplot
